Taskfile.yml

```diff
@@ -60,7 +60,9 @@ tasks:
       # Source: https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/github-workflow.json
       WORKFLOW_SCHEMA_URL: https://json.schemastore.org/github-workflow
       WORKFLOW_SCHEMA_PATH:
-        sh: mktemp -t workflow-schema-XXXXXXXXXX.json
+        sh: |
+          RAW_PATH="$(mktemp -t workflow-schema-XXXXXXXXXX.json)"
+          task utility:normalize-path RAW_PATH="$RAW_PATH"
```

Comment:
I considered it but was kind of indecisive because I already need to do the same for creating temporary folders for another upcoming change (workaround for a bug in npx).

So I would either need something like this:

```yaml
version: "3"
tasks:
  foo:
    vars:
      SOME_TEMP_FILE:
        sh: task utility:mktempfile TEMPLATE=some-temp-file-XXXXXXXXXX.json
      SOME_TEMP_DIR:
        sh: task utility:mktempdir TEMPLATE=some-temp-dir-XXXXXXXXXX
    cmds:
      - echo "{{.SOME_TEMP_FILE}}"
      - echo "{{.SOME_TEMP_DIR}}"

  utility:mktempfile:
    vars:
      RAW_PATH:
        sh: mktemp -t {{.TEMPLATE}}
    cmds:
      - task: utility:normalize-path
        vars:
          RAW_PATH: "{{.RAW_PATH}}"


  utility:mktempdir:
    vars:
      RAW_PATH:
        sh: mktemp --directory --tmpdir {{.TEMPLATE}}
    cmds:
      - task: utility:normalize-path
        vars:
          RAW_PATH: "{{.RAW_PATH}}"

  utility:normalize-path:
    desc: Print a normalized version of the path passed via the RAW_PATH variable to stdout
    cmds:
      - |
        if [[ "{{.OS}}" == "Windows_NT" ]] && which cygpath &>/dev/null; then
            # Even though the shell handles POSIX format absolute paths as expected, external applications do not.
            # So paths passed to such applications must first be converted to Windows format.
            cygpath -w "{{.RAW_PATH}}"
        else
          echo "{{.RAW_PATH}}"
        fi
```

or this:

```yaml
version: "3"
tasks:
  foo:
    vars:
      SOME_TEMP_FILE:
        sh: task utility:mktemp TEMPLATE=some-temp-file-XXXXXXXXXX.json TYPE=file
      SOME_TEMP_DIR:
        sh: task utility:mktemp TEMPLATE=some-temp-dir-XXXXXXXXXX TYPE=dir
    cmds:
      - echo "{{.SOME_TEMP_FILE}}"
      - echo "{{.SOME_TEMP_DIR}}"

  utility:mktemp:
    vars:
      RAW_PATH:
        sh: |
          if [[ "{{.TYPE}}" == "file" ]]; then
            mktemp -t {{.TEMPLATE}}
          else
            mktemp --directory --tmpdir {{.TEMPLATE}}
          fi
    cmds:
      - task: utility:normalize-path
        vars:
          RAW_PATH: "{{.RAW_PATH}}"

  utility:normalize-path:
    desc: Print a normalized version of the path passed via the RAW_PATH variable to stdout
    cmds:
      - |
        if [[ "{{.OS}}" == "Windows_NT" ]] && which cygpath &>/dev/null; then
            # Even though the shell handles POSIX format absolute paths as expected, external applications do not.
            # So paths passed to such applications must first be converted to Windows format.
            cygpath -w "{{.RAW_PATH}}"
        else
          echo "{{.RAW_PATH}}"
        fi
```

Due to giving a lot of attention to thoroughly validating the assets here at the source, this repository has a lot more of this stuff in the taskfile than a regular project repo will. The two distributable taskfiles I applied it to are not quite so commonly used in other projects.

Do you think it is worth the tradeoff between complexity vs code duplication to do it this way @cmaglie?

Reply:
I prefer the first solution, with `mktempfile` and `mktempdir`.

> Do you think it is worth the tradeoff between complexity vs code duplication to do it this way @cmaglie?

Yes. The rationale is that these functions should help you quickly do the most repetitive task. In this case, the most repetitive task is making a temp file/folder, not normalizing a path, that's why I'd like to see the full mktemp action covered by the task target.

Moreover, having a "normalize path" action poses the question of **when** you should normalize the path? Why only `mktemp` and not other paths?

Reply:
> I prefer the first solution

I have now done it:
https://github.com/arduino/tooling-project-assets/compare/c67befe4f47928d99591b29e15f384bf6c246a14..57141c4224d06e37ca773565ba1cf4e85a7b35b1

>these functions should help you quickly do the most repetitive task

These are only intended to be used by other tasks to avoid code duplication. They are not intended to be used directly by the user. So "quickly" is not applicable. But "repetitive" might describe the many places they are called by other tasks.

>when you should normalize the path?

The issue can only happen with absolute paths. Even with absolute paths, if they are used only by the integrated shell interpreter or by the external POSIX/Bash shell's utilities, and even by some standalone applications there is no problem. But probably the safest thing is to just normalize every absolute path.

>Why only mktemp

Because, in the way it is used here, it returns absolute paths (`/tmp/<TEMPLATE>`).

>not other paths?

They are relative.

Reply:
> These are only intended to be used by other tasks to avoid code duplication. They are not intended to be used directly by the user. So "quickly" is not applicable. But "repetitive" might describe the many places they are called by other tasks.

Yes maybe "quickly" is not the best word to describe my thought. What I mean is that I don't like the added cognitive load. When reading the script:
```
          RAW_PATH="$(mktemp -t workflow-schema-XXXXXXXXXX.json)"
          task utility:normalize-path RAW_PATH="$RAW_PATH"
```
is not immediately clear what these two lines are doing, you need to dig inside the evaluation operator `$(....)` to actually understand that they are making a temp file. And this is general, even if we use these scripts internally.

There are other advantages too besides the removed code duplication, for example, if we will ever need to change the way temp folders are created we are able to change it in one place to have the fix propagated, etc.